backend/EasySpeak.Core/EasySpeak.Core.BLL/Services/LessonSchedulerService.cs

```diff
@@ -0,0 +1,195 @@
+using System.Data.Entity;
+using AutoMapper;
+using EasySpeak.Core.BLL.Options;
+using EasySpeak.Core.Common.DTO.Notification;
+using EasySpeak.Core.Common.Enums;
+using EasySpeak.Core.DAL.Context;
+using EasySpeak.Core.DAL.Entities;
+using EasySpeak.RabbitMQ.Interfaces;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Hosting;
+using Microsoft.Extensions.Options;
+using EntityFrameworkQueryableExtensions = Microsoft.EntityFrameworkCore.EntityFrameworkQueryableExtensions;
+
+namespace EasySpeak.Core.BLL.Services;
+
+public class LessonSchedulerService : BackgroundService
+{
+    private readonly LessonSchedulerOptions _schedulerOptions;
+    private readonly IServiceProvider _services;
+    private readonly IMessageProducer _messageProducer;
+    private readonly IMapper _mapper;
+    private readonly PeriodicTimer _periodicTimer;
+
+    public LessonSchedulerService(IServiceProvider services, IMessageProducer messageProducer,
+        IOptions<LessonSchedulerOptions> schedulerOptions, IMapper mapper)
+    {
+        _schedulerOptions = schedulerOptions.Value;
+        _services = services;
+        _messageProducer = messageProducer;
+        _mapper = mapper;
+        _periodicTimer = new PeriodicTimer(TimeSpan.FromMinutes(_schedulerOptions.CheckPeriod));
+    }
+    
+    protected override async Task ExecuteAsync(CancellationToken cancellationToken)
+    {
+        while (await _periodicTimer.WaitForNextTickAsync(cancellationToken) && 
+               !cancellationToken.IsCancellationRequested)
+        {
+            var lessons = await GetLessons(cancellationToken);
+
+            var pendingTasks = lessons.Select(ScheduleSubscribersNotification);
+
+            await Task.WhenAll(pendingTasks);
+        }
+    }
+
+    private async Task<List<LessonDelayDto>> GetLessons(CancellationToken cancellationToken)
+    {
+        var currentTime = DateTime.UtcNow;
+
+        var fromTime = currentTime.AddMinutes(_schedulerOptions.CheckPeriod);
+
+        var toTime = currentTime.AddMinutes(_schedulerOptions.CheckPeriod * 2);
+
+        using var scope = _services.CreateScope();
+        
+        var context = scope.ServiceProvider.GetRequiredService<EasySpeakCoreContext>();
```

Comment:
DI?

Reply:
It's background service, so it's not possible to get this service via usual DI